rust/main/chains/hyperlane-starknet/src/mailbox.rs

```diff
@@ -0,0 +1,258 @@
+        locator: &ContractLocator,
+        signer: Signer,
+    ) -> ChainResult<Self> {
+        let account = build_single_owner_account(
+            &conn.url,
+            signer.local_wallet(),
+            &signer.address,
+            false,
+            locator.domain.id(),
+        );
+
+        let contract = StarknetMailboxInternal::new(
+            locator
+                .address
+                .try_into()
+                .map_err(HyperlaneStarknetError::BytesConversionError)?,
+            account,
+        );
+
+        Ok(Self {
+            contract: Arc::new(contract),
+            provider: StarknetProvider::new(locator.domain.clone(), conn),
+            conn: conn.clone(),
+        })
+    }
+
+    /// Returns a ContractCall that processes the provided message.
+    /// If the provided tx_gas_limit is None, gas estimation occurs.
+    async fn process_contract_call(
+        &self,
+        message: &HyperlaneMessage,
+        metadata: &[u8],
+        tx_gas_estimate: Option<U256>,
+    ) -> ChainResult<Execution<'_, SingleOwnerAccount<AnyProvider, LocalWallet>>> {
+        let tx = self.contract.process(
+            &to_mailbox_bytes(metadata),
+            &StarknetMessage {
+                version: message.version,
+                nonce: message.nonce,
+                origin: message.origin,
+                sender: StarknetU256::from_bytes_be(&message.sender.to_fixed_bytes()),
+                destination: message.destination,
+                recipient: StarknetU256::from_bytes_be(&message.recipient.to_fixed_bytes()),
+                body: to_mailbox_bytes(&message.body),
+            },
+        );
+
+        let gas_estimate = match tx_gas_estimate {
+            Some(estimate) => estimate
+                .try_into()
+                .map_err(Into::<HyperlaneStarknetError>::into)?,
+            None => {
+                tx.estimate_fee()
+                    .await
+                    .map_err(|e| HyperlaneStarknetError::AccountError(e.to_string()))?
+                    .overall_fee
+            }
+        };
+        Ok(tx.max_fee(gas_estimate * FieldElement::TWO))
+    }
+
+    #[allow(unused)]
+    pub fn contract(
+        &self,
+    ) -> &StarknetMailboxInternal<SingleOwnerAccount<AnyProvider, LocalWallet>> {
+        &self.contract
+    }
+}
+
+impl HyperlaneChain for StarknetMailbox {
+    fn domain(&self) -> &HyperlaneDomain {
+        &self.provider.domain()
+    }
+
+    fn provider(&self) -> Box<dyn HyperlaneProvider> {
+        Box::new(self.provider.clone())
+    }
+}
+
+impl HyperlaneContract for StarknetMailbox {
+    fn address(&self) -> H256 {
+        self.contract.address.into()
+    }
+}
+
+#[async_trait]
+impl Mailbox for StarknetMailbox {
+    #[instrument(skip(self))]
+    async fn count(&self, reorg_period: &ReorgPeriod) -> ChainResult<u32> {
+        let block_number =
+            get_block_height_for_reorg_period(&self.provider.rpc_client(), reorg_period).await?;
```

Comment:
nit: could you refactor this function to return the latest block number if `reorg_period` is None? This way you'd always have a `BlockId::Number(b)` too call the contract with and can simplify the code below

Reply:
done

Reply:
nice, I also meant that inside `get_block_height_for_reorg_period` you could modify the `ReorgPeriod::None` arm to be
```rust
    let block_height = match reorg_period {
        ReorgPeriod::Blocks(blocks) => {
            let tip = provider.latest_block_height().await?;
            let block_height = tip - blocks.get() as u64;
            Some(block_height)
        }
        ReorgPeriod::None => provider.latest_block_height().await?;
```

since this doesn't affect the other callers of this function AFAICT, and the signature would change to this, to no longer return an `Option`:
```
pub(crate) async fn get_block_height_for_reorg_period(
    provider: &WasmGrpcProvider,
    reorg_period: &ReorgPeriod,
) -> ChainResult<u64> {
```

Reply:
changed!